### PR TITLE
[rfr] More User and Node porting

### PR DESCRIPTION
**Pull request by sloria**
- `User.update_affiliated_institutions_based_on_email_domain`
- Change default=datetime.utcnow -> default=timezone.now
- `Node#add_unregistered_contributor` and related User methods
- `User#get_projects_in_common` and `User#n_projects_in_common`

### CODE REVIEW

File: osf_models/models/node.py [hunk 1]
```diff
@@ -213,7 +235,12 @@ def get_permissions(self, user):
         return perm
 
     def has_permission(self, user, permission):
-        return getattr(user.contributor_set.get(node=self), permission, False)
+        try:
+            contrib = user.contributor_set.get(node=self)
+        except Contributor.DoesNotExist:
+            return False
+        else:
+            return getattr(contrib, permission, False)
```
Comment (cwisecarver) [inline osf_models/models/node.py]: Good catch.

File: osf_models/models/user.py [hunk 2]
```diff
@@ -816,3 +822,100 @@ def add_system_tag(self, tag):
...
+        """Returns either a collection of "shared projects" (projects that both users are contributors for)
+        or just their primary keys
+        """
+        if primary_keys:
+            projects_contributed_to = set(self.contributed.values_list('_guid__guid', flat=True))
+            other_projects_primary_keys = set(other_user.contributed.values_list('_guid__guid', flat=True))
+            return projects_contributed_to.intersection(other_projects_primary_keys)
+        else:
+            projects_contributed_to = set(self.contributed)
+            return projects_contributed_to.intersection(other_user.contributed)
```
Comment (cwisecarver) [inline osf_models/models/user.py]: You should be able to do `Node.objects.filter(contributors=self).filter(contributors=other_user).values_list('_guid__guid', flat=True)` and
`Node.objects.filter(contributors=self).filter(contributors=other_user)` respectively. That will get them down to one query each. It's not particularly obvious that you can do that with the orm, I knew you could. I just didn't remember how. Took a  good bit of googling. `filter(contributors__in=[list, of, users])` would match ANY not ALL of them.

File: osf_models/models/user.py [hunk 3]
```diff
@@ -816,3 +822,100 @@ def add_system_tag(self, tag):
...
+        :raises: ValueError if a record doesn't exist for the given project ID
+        :rtype: dict
+        :returns: The unclaimed record for the project
+        """
+        config = apps.get_app_config('osf_models')
+        uid = self._primary_key
+        base_url = config.domain if external else '/'
+        unclaimed_record = self.get_unclaimed_record(project_id)
+        token = unclaimed_record['token']
+        return '{base_url}user/{uid}/{project_id}/claim/?token={token}'\
```
Comment (cwisecarver) [inline osf_models/models/user.py]: The django way: https://docs.djangoproject.com/en/1.10/ref/request-response/#django.http.HttpRequest.build_absolute_uri

File: osf_models/models/user.py [hunk 4]
```diff
@@ -816,3 +822,100 @@ def add_system_tag(self, tag):
...
+
+    def update_affiliated_institutions_by_email_domain(self):
+        """
+        Append affiliated_institutions by email domain.
+        :return:
+        """
+        Institution = apps.get_model('osf_models.Institution')
+        try:
+            email_domains = [email.split('@')[1].lower() for email in self.emails]
+            insts = Institution.find(Q('email_domains', 'overlap', email_domains))
```
Comment (cwisecarver) [inline osf_models/models/user.py]: Do we want to standardize on using either django or modm on stuff we're rewriting? I've been going back and forth but trying to switch things to django ORM methods when I can. We're going to have to go back through everything anyway, I just wondered.

File: osf_models/models/user.py [hunk 5]
```diff
@@ -816,3 +822,100 @@ def add_system_tag(self, tag):
...
+            for inst in insts:
+                if inst not in self._affiliated_institutions.all():
```
Comment (cwisecarver) [inline osf_models/models/user.py]: You can add multiple at once in one query.
`self._affiliated_institutions.add(*list_of_new_insts)`

I also think there's a way to do this whole thing in one query. I just haven't figured it out yet. I'm going to move on for now though.

File: osf_models/models/user.py (hunk 4 above)
Comment (sloria) [inline osf_models/models/user.py]: I don't feel strongly either way. Here I just decided to make the minimal change for the query to work. For queries that require more work to port, I'll use the Django API.

Comment (cwisecarver) [inline osf_models/models/user.py]: 👍

**Comment by cwisecarver** [comment]
Other than a couple of minor django things lgtm. They aren't necessary. Just either less queries or less code.

File: osf_models/models/user.py (hunk 3 above)
Comment (sloria) [inline osf_models/models/user.py]: `build_absolute_uri` is an instance method of a Django HttpRequest, which we don't have access to here.

Comment (cwisecarver) [inline osf_models/models/user.py]: :(

**Comment by sloria** [comment]
Made minor refactors, rebased, and updated initial_migration.

**Pull request merged**